### ISSUE TITLE
Freezing a BatchNormalization layer by setting trainable = false does not work. Weights are still being updated.  

### ISSUE THREAD

**Issue by afourast** [keras-team/keras#7085]
The issue seems to be that that the [updates that are created](https://github.com/fchollet/keras/blob/de73eda89a916c4dd46ce74058bb2664455ed9db/keras/layers/normalization.py#L179) on applying the BatchNorm layer are [added to the train function](https://github.com/fchollet/keras/blob/de73eda89a916c4dd46ce74058bb2664455ed9db/keras/engine/training.py#L924) even when they act on non-trainable weights. 

gist to code:
https://gist.github.com/afourast/0d7545174c1b8fb7b0f82d7efbf31743

Please make sure that the boxes below are checked before you submit your issue. If your issue is an implementation question, please ask your question on [StackOverflow](http://stackoverflow.com/questions/tagged/keras) or [join the Keras Slack channel](https://keras-slack-autojoin.herokuapp.com/) and ask there instead of filing a GitHub issue.

Thank you!

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [x] If running on TensorFlow, check that you are up-to-date with the latest version. The installation instructions can be found [here](https://www.tensorflow.org/get_started/os_setup).

- [ ] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [x] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by Danielhiversen** [comment]
https://github.com/fchollet/keras/issues/7051#issuecomment-309851005

**Comment by hanzy123** [comment]
BN layer has alpha,beta,varience and mean layers, the alpha and beta layers' param can change the 
trainable attribute,but the varience and mean param can not set untrainable. Although you set the BN layer trainable = False ,it will not work

**Comment by ahundt** [comment]
Locking BN is a technique used in segmentation so I think this is an option we need to add if it does not currently exist

**Comment by nsmetanin** [comment]
Is there any workaround available for this issue? It turns into hell when you're trying to use single shared model body (pre-trained model with trainable=False) inside multiple models which are trained on different datasets

**Comment by afourast** [comment]
If you are willing to give up on the Keras training routines (fit, fit_batch, fit_generator etc), you can define a custom training function in the way that Keras does for them (https://github.com/fchollet/keras/blob/d3db58c8bf1ef9d078b3cf5d828d22346df6469b/keras/engine/training.py#L948), where you won't include the BN statistics updates. For example:

https://gist.github.com/afourast/018722309ac2a272ce4985190ec52742

In my models there are no other updates added to the 'updates' list besides the BN statistics, but you should check that with your models. Also I have only tested this with a tf backend. You will  have to define a similar function for testing.

I've personally found that using a custom training function convenient since you can also do things like use tf queues and add tensorboard summaries.

Hope that helps

**Comment by nsmetanin** [comment]
Thanks, but it seems too tricky for my case. 

I've found another way to workaround that: you can set layer._per_input_updates = {} on your batch norm layers which should be fixed during training. It actually works, these layer weights stay the same, but it's still looks like dirty hack.

**Comment by NakramR** [comment]
I'm running into the same issue, is there a technical reason it can't be fixed?

**Comment by hanzy123** [comment]
U can find:BatchNormalization layer was combined by four small layers:alpha layer,beta layer,and two other weight-layers,your `Trainable=False` order just keep two other weight-layers unchanged

**Comment by NakramR** [comment]
Right, and can the alpha and beta layers be frozen, and if so how?

**Comment by LazyMammal** [comment]
Thank you @nsmetanin for the suggestion.  This worked to freeze batch norm layers.

```
for layer in model.layers:
    layer.trainable = False
    if isinstance(layer, keras.layers.normalization.BatchNormalization):
        layer._per_input_updates = {}
```

**Comment by ahundt** [comment]
Freezing layers is now available in the most recent release, simply set trainable=False for the batchnorm layers.

https://github.com/keras-team/keras/releases/tag/2.1.3

**Comment by ViaFerrata** [comment]
Is there any way to get back the old behaviour? 
I'd like to get the trainable weights frozen and the non-trainable weights unfrozen, as before.

**Comment by ahundt** [comment]
![broken workflow](https://imgs.xkcd.com/comics/workflow.png)
@ViaFerrata I don't have a real answer other than look back in the commit history to see how it was done before.

**Comment by fchollet** [comment]
You can set layer.stateful = True on your BN layer to get this behavior.

**Comment by ViaFerrata** [comment]
Thanks a lot for the quick answer, I will try that :)

**Comment by ozabluda** [comment]
@ViaFerrata, just checking, what do you want the old behavior for?

**Comment by ozabluda** [comment]
Training never uses anything other than current batch for training, but it updates the running averages for inference. If training is resumed, I think running average is reset to zero.

**Comment by ViaFerrata** [comment]
Thank you for the explanation, I've misread that in the paper. Then it makes sense that I get worse results with stateful=False during retraining.

**Comment by zhimengfan1990** [comment]
From the code of Keras 2.1.3, what I see is `training=False` rather than `trainable=False`
https://github.com/keras-team/keras/blob/2.1.3/keras/layers/normalization.py#L175

**Comment by udevnl** [comment]
Does you model contain a Upsampling2D layer AND did you specify interpolation?
```
# Example
x = UpSampling2D(2, interpolation="bilinear")(x)
```

It seems the setting 'interpolation' is NOT saved with the model. Therefore, when loading the model, UpSampling2D layers get the default 'nearest' interpolation when loading. Because the model was trained with 'bilinear' interpolation it will not perform as well when given 'nearest' interpolation after loading.

In tensorflow 1.14.0 this seems to be the case. It seems fixed in tensorflow 1.15.0.

WORKAROUND 1
1.  Create the model with the original code
2.  Now load the save weights from either the .h5 of .hdf5 file using model.load_weights(...)

WORKAROUND 2
1.  In the model JSON file, add *"interpolation": "bilinear"* to the config of UpSampling2D layers.
2.  Now load, this works, even in tensorflow 1.14.0

WORKAROUND 3
1.  After loading the model, look in model.layers and for each UpSampling2D layer, change the interpolation


I also noticed that the .h5 file is smaller (8MB) than the .hdf5 file (25MB) but the contents seems similar, probably compression. I proved this by getting original model performance after loading either. Also comparing the loaded weights and precision (float32) was equal.

Lastly, took me 4 hours to find. I found this issue by having the good and bad model simultaneously loaded in a Jupyter notebook. Then, I let both models predict a single sample and compared the output after each layer until a difference occurred. In my case after layer 35 which was the first Upsampling2D layer.


```
layer_no = 34
inf_good = Model(inputs=model_good.input, outputs=[model_good.layers[layer_no].output])
inf_bad = Model(inputs=model_bad.input, outputs=[model_bad.layers[layer_no].output])

pred_sample = 2
plt.imshow(np.squeeze(X[pred_sample]))

p_good = inf_good.predict(X[pred_sample:pred_sample+1])
p_bad = inf_bad.predict(X[pred_sample:pred_sample+1])

(p_good == p_bad).all()
```

Hope this helps
